components/fxa-client/src/pairing_channel/client_states.rs

```diff
@@ -0,0 +1,388 @@
+        self.record_layer.flush()?;
+        self.session_id = session_id;
+        Ok(self)
+    }
+
+    fn recv_handshake_message(mut self, message: HandshakeMessage) -> Result<ClientWaitEe<F>> {
+        if *message.get_type() != HandshakeType::ServerHello {
+            anyhow::bail!("Invalid message");
+        }
+        if hex::encode(&self.session_id) != hex::encode(message.get_session_id()) {
+            anyhow::bail!("Invalid session id")
+        }
+        let psk_ext = message
+            .get_extensions()
+            .iter()
+            .find(|ext| ext.get_type_tag() == PRE_SHARED_KEY_ID)
+            .ok_or_else(|| anyhow::Error::msg("No psk extension!"))?;
+        if psk_ext.get_selected_identity()? != 0 {
+            anyhow::bail!("Invalid selected identity")
+        }
+        self.key_schedule.add_ecdhe(b"")?;
+        let send_key = self.key_schedule.get_client_handshake_secret()?;
+        let recv_key = self.key_schedule.get_server_handshake_secret()?;
+        self.record_layer.set_send_key(&send_key)?;
+        self.record_layer.set_recv_key(&recv_key)?;
+        let client_wait_ee: ClientWaitEe<F> = self.into();
+        Ok(client_wait_ee.initialize()?)
+    }
+}
+
+pub struct ClientWaitEe<F: FnMut(&[u8]) -> Result<()>> {
+    key_schedule: KeySchedule,
+    record_layer: RecordLayer<F>,
+    has_seen_change_cipher_spec: bool,
+}
+impl<F: FnMut(&[u8]) -> Result<()>> From<ClientWaitSh<F>> for ClientWaitEe<F> {
+    fn from(client_wait_sh: ClientWaitSh<F>) -> Self {
+        Self {
+            key_schedule: client_wait_sh.key_schedule,
+            record_layer: client_wait_sh.record_layer,
+            has_seen_change_cipher_spec: false,
+        }
+    }
+}
+
+impl<F: FnMut(&[u8]) -> Result<()>> ClientWaitEe<F> {
+    fn initialize(mut self) -> Result<Self> {
+        self.record_layer.flush()?;
+        Ok(self)
+    }
+    fn recv_change_cipher_spec(&mut self, bytes: &[u8]) -> Result<()> {
+        if self.has_seen_change_cipher_spec {
+            anyhow::bail!("Seen change cipher spec before!")
+        }
+        if bytes.len() != 1 || bytes[0] != 1 {
+            anyhow::bail!("Invalid bytes fo change cipher spec")
+        }
+        self.has_seen_change_cipher_spec = true;
+        Ok(())
+    }
+
+    fn recv_handshake_message(self, message: HandshakeMessage) -> Result<ClientWaitFinished<F>> {
+        // We don't make use of any encrypted extensions, but we still
+        // have to wait for the server to send the (empty) list of them.
+        if *message.get_type() != HandshakeType::EncryptedExtensions {
+            anyhow::bail!("invalid message");
+        }
+        // We do not support any EncryptedExtensions.
+        if !message.get_extensions().is_empty() {
+            anyhow::bail!("Unsupported extension")
+        }
+        let server_finished_transcript = self.key_schedule.get_transcript();
+        let client_wait_finished: ClientWaitFinished<F> = self.into();
```

Comment:
This use of `From` is a neat trick! I think I like it, although it's probably worth a bit of scene-setting at the top of the file to help the reader know what to expect.

I wonder if this would read more clearly as `ClientWaitFinished::from(self)` than `self.into()`, may be worth trying to see what you think (and I really do mean "I wonder", not "I think this but want to sound modest about my own suggestions", just to be clear 😆)

Reply:
Agree, sounds clearer! And ya the `From` trick allows us to ensure we're not doing any invalid transitions at compile time 😄